akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala

```diff
@@ -0,0 +1,291 @@
+          resp.headers.find(_.is("location")).flatMap(location => {
+            val item = if (sameOrigin(req.uri, Uri(location.value))) s.sameOrigin else s.crossOrigin
+            if (item.getAllow) {
+              val headers = item.getForwardHeaders match {
+                case All => req.headers
+                case Zero => Nil
+                case Only(hs) => req.headers.filter(h => hs.contains(h.name))
+              }
+              Some(req.withUri(location.value).withHeaders(headers).withMethod(method))
+            } else {
+              None
+            }
+          })
+        }
+      } else {
+        None
+      }
+    }
+
+  def sameOrigin(u1: Uri, u2: Uri): Boolean = {
+    if (u2.isRelative) true
+    else if (u1.isAbsolute && u2.isAbsolute && u1.authority == u2.authority) true
+    else false
+  }
+
+  trait OpOutPrinter[F, _ <: Outlet[F]] {
+    def print(t: F): Unit
+  }
+
+  trait OpInPrinter[T <: Inlet[_]] {
+    def print(): Unit
+  }
+
+  def printy(s: Any) = ()//println(s)
+
+  implicit val RequestOutPrinterReq = new OpOutPrinter[HttpRequest, Outlet[HttpRequest]] {
+    override def print(t: HttpRequest): Unit = printy(s"\tpushing $t to requestOut")
+  }
+  implicit val ResponseOutPrinterResp = new OpOutPrinter[HttpResponse, Outlet[HttpResponse]] {
+    override def print(t: HttpResponse): Unit = printy(s"\tpushing $t to responseOut")
+  }
+  implicit val RequestInPrinterReq = new OpInPrinter[Inlet[HttpRequest]] {
+    override def print(): Unit = {
+      printy("\tpulling from requestIn")
+    }
+  }
+  implicit val ResponseInPrinterResp = new OpInPrinter[Inlet[HttpResponse]] {
+    override def print(): Unit = {
+      printy("\tpulling from responseIn")
+    }
+  }
+
+  def apply(settings: ClientAutoRedirectSettings) = BidiFlow.fromGraph(new RedirectSupportStage(settings, defaultRedirectMapper))
+
+  def apply(settings: ClientAutoRedirectSettings, redirectMapper: RedirectMapper) = BidiFlow.fromGraph(new RedirectSupportStage(settings, redirectMapper))
+}
+
+/**
+  * A BidiShape that sits on top of OutgoingConnectionBlueprint and routes the redirect responses back to the
+  * downstream outgoing connection if [[redirectMapper]] allows, or upstream otherwise
+  * @param settings Redirect configuration
+  * @param redirectMapper returns Some(HttpRequest) with the redirect HttpRequest if automatic redirect must be
+  *                       performed, or None if received response must go directly upstream
+  */
+class RedirectSupportStage(settings: ClientAutoRedirectSettings, redirectMapper: RedirectMapper) extends GraphStage[BidiShape[HttpRequest, HttpRequest, HttpResponse, HttpResponse]] {
+  /**
+    * in1 is where the original messages come from
+    */
+  val requestIn: Inlet[HttpRequest] = Inlet("Input1")
+
+  /**
+    * in2 is the responses to original messages. If it's a redirect, its not forwarded upstream, but goes downstream
+    */
+  val responseIn: Inlet[HttpResponse] = Inlet("Input2")
```

Comment:
(If something fails, the name is only printed - so it's more helpful to include names like this, we're trying to standardise on this pattern now)